### PR TITLE
Add AUC ROC bootstrapping QC comparison

### PR DESCRIPTION
**Pull request by jenna-tomkinson**
# Add AUC ROC bootstrapping QC comparison

In this PR, we add code to perform a bootstrapping method that applies the model (QC trained and no QC trained) onto a holdout dataset that has been QC'd to evaluate how the ROC AUC changes across different subsamples.

### CODE REVIEW

File: 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# ## Perform bootstrapping ROC AUC method to determine if QC significantly improves performance of model
```
Comment (gwaybio) [inline 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py:4]: We will benefit in the future from doing one of two things (maybe both).

1. Clarify what you mean by "new" in the docstring - what precisely is different?
2. Change the name of the file to be more descriptive of what you mean by "new"

**Review by gwaybio — APPROVED** [review]
This looks good! I also wonder at the size of the Git LFS data - these are relatively low value right? If they're small, could we just add them to the repo? If they're medium then prob ok to keep as LFS, but if large, then maybe best to move to figshare. Interested in hearing your thoughts!

File: 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_comparisons.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    bootstrapped_scores : np.ndarray
+        An array of bootstrapped ROC AUC scores. Each element represents the 
+        ROC AUC computed for a resampled dataset.
+    """
+
+    bootstrapped_scores = []
+    for i in range(n_bootstraps):
+        indices = resample(np.arange(len(y_true)), replace=True)
+        if len(np.unique(y_true[indices])) < 2:
+            continue
```
Comment (gwaybio) [inline 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_comparisons.py:71]: I'm not sure if this if statement actually is doing anything. There's no `else` part to direct the logic

File: 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py (hunk 1 above)
Comment (gwaybio) [reply 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py:4]: I'm not sure if you addresed this?

File: 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+from training_utils import get_X_y_data
+
+
+# ## Define function
+
+# In[2]:
+
+
+# Define function to preform bootstrapping
+def bootstrap_roc_auc(y_true, y_pred, n_bootstraps=1000):
```
Comment (gwaybio) [inline 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py:41]: probably best to not define this twice but import it instead

File: 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+plt.legend(loc="upper left", fontsize=10)
+plt.xlabel("ROC AUC Score", fontsize=12)
+plt.ylabel("Frequency", fontsize=12)
+plt.title("Bootstrap ROC AUC Distributions", fontsize=14, fontweight="bold")
+plt.grid(True, linestyle="--", alpha=0.6)
+plt.tight_layout()
+
+# save figure
+plt.savefig("./figures/bootstrap_plot_holdout_each_model.png", dpi=600)
```
Comment (gwaybio) [inline 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py:242]: is it possible to be a bit more specific about what this figure represents? What does "holdout_each_model" mean?

File: 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    np.mean(scores_model2),
+    color=darker_coral,
+    linestyle="dashed",
+    linewidth=2,
+)
+
+plt.legend(loc="upper left", fontsize=10)
+plt.xlabel("ROC AUC Score", fontsize=12)
+plt.ylabel("Frequency", fontsize=12)
+plt.title("Bootstrap ROC AUC Distributions", fontsize=14, fontweight="bold")
```
Comment (gwaybio) [inline 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py:237]: You can also consider adding the nuance difference of this approach to the title

**Review by gwaybio — APPROVED** [review]

File: 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py (hunk 1 above)
Comment (jenna-tomkinson) [reply 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py:4]: I hadn't addressed any of the comments yet, haha. I was looking to chat IRL but I will try and address and we can chat later.

Comment (jenna-tomkinson) [reply 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py:4]: Done! Let me know your thoughts on the changes!

File: 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py (hunk 3 above)
Comment (jenna-tomkinson) [reply 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py:41]: Agreed, adding utility now!

File: 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py (hunk 4 above)
Comment (jenna-tomkinson) [reply 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py:242]: Updated these now!

File: 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py (hunk 5 above)
Comment (jenna-tomkinson) [reply 5.machine_learning/1.evaluate_models/scripts/3.ROC_QC_new_method.py:237]: Will do!

File: 3.process_cfret_features/scripts/2.single_cell_processing.py [hunk 6]
```diff
@@ -18,18 +18,25 @@
 
 # ## Set paths and variables
 
-# In[2]:
+# In[ ]:
```
Comment (gwaybio) [inline 3.process_cfret_features/scripts/2.single_cell_processing.py]: should be executed?

**Review by gwaybio — APPROVED** [review]

**Comment by jenna-tomkinson** [comment]
Thanks for the review @gwaybio! I am merging now!

**Pull request merged**